src/Shared/UrlDecoder/UrlDecoder.cs

```diff
@@ -345,5 +344,296 @@ private static bool SkipUnescape(int value1, int value2, bool isFormEncoding)
+                    }
+
+                    sourceIndex = decodeIndex;
+                }
+                else
+                {
+                    buffer[destinationIndex++] = buffer[sourceIndex++];
+                }
+            }
+
+            return destinationIndex;
+        }
+
+        /// <summary>
+        /// Unescape the percent-encodings
+        /// </summary>
+        /// <param name="sourceIndex">The iterator point to the first % char</param>
+        /// <param name="destinationIndex">The place to write to</param>
+        /// <param name="buffer">The char array</param>
+        private static bool DecodeCore(ref int sourceIndex, ref int destinationIndex, Span<char> buffer)
+        {
+            // preserves the original head. if the percent-encodings cannot be interpreted as sequence of UTF-8 octets,
+            // chars from this till the last scanned one will be copied to the memory pointed by writer.
+            var char1 = UnescapePercentEncoding(ref sourceIndex, buffer);
+            if (char1 == -1)
+            {
+                return false;
+            }
+
+            if (char1 == 0)
+            {
+                throw new InvalidOperationException("The path contains null characters.");
+            }
+
+            if (char1 <= 0x7F)
+            {
+                // first char < U+007f, it is a single char ASCII
+                buffer[destinationIndex++] = (char)char1;
+                return true;
+            }
+
+            // anticipate more chars
+            var currentDecodeBits = 0;
+            var charCount = 1;
+            var expectValueMin = 0;
+            if ((char1 & 0xE0) == 0xC0)
+            {
+                // 110x xxxx, expect one more char
+                currentDecodeBits = char1 & 0x1F;
+                charCount = 2;
+                expectValueMin = 0x80;
+            }
+            else if ((char1 & 0xF0) == 0xE0)
+            {
+                // 1110 xxxx, expect two more chars
+                currentDecodeBits = char1 & 0x0F;
+                charCount = 3;
+                expectValueMin = 0x800;
+            }
+            else if ((char1 & 0xF8) == 0xF0)
+            {
+                // 1111 0xxx, expect three more chars
+                currentDecodeBits = char1 & 0x07;
+                charCount = 4;
+                expectValueMin = 0x10000;
+            }
+            else
+            {
+                // invalid first char
+                return false;
+            }
+
+            var remainingChars = charCount - 1;
+            while (remainingChars > 0)
+            {
+                // read following three chars
+                if (sourceIndex == buffer.Length)
+                {
+                    return false;
+                }
+
+                var nextSourceIndex = sourceIndex;
+                var nextChar = UnescapePercentEncoding(ref nextSourceIndex, buffer);
+                if (nextChar == -1)
```

Comment:
If you're _reeeeally_ trying to squeeze every last drop of perf from this routine, you could drop this check, as it's already covered by the `if ((nextChar & 0xC0) != 0x80)` line below. (An input value of -1 would fail that check.)

Reply:
Good catch. I will also add a code comment, so it is clear for the reader.